### ISSUE TITLE
API does not always reconnect after disconnect.

### ISSUE THREAD

**Issue by igoberman** [nats-io/java-nats#231]
We have application using nats server running on local host. We use latest jnats 2.4.3. Once in a while, under heavy load, nats server closes connection. This causes 'connection reset' on the client side. However, jnats API does not reconnect. This is a severe problem causing our application to stop working.

Last exception was CancellationException related to cleanUpPongQueue. NATS never issued reconnect attempts after that.
I have reported this before in https://github.com/nats-io/java-nats/issues/226, but moving it to a separate thread as it may be different.

2019-04-17T09:30:12.346 [pool-4-thread-1] ERROR c.b.o.c.nats.NatsConnectionProvider - Nats 'nats://127.0.0.1:4222' exception: {}
java.net.SocketException: Connection reset
at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115)
at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
at io.nats.client.impl.SocketDataPort.write(SocketDataPort.java:104)
at io.nats.client.impl.NatsConnectionWriter.run(NatsConnectionWriter.java:160)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
2019-04-17T09:30:12.411 [pool-4-thread-1] ERROR c.b.o.c.nats.NatsConnectionProvider - Nats 'null' exception: {}
java.net.SocketException: Connection reset
at java.net.SocketInputStream.read(SocketInputStream.java:210)
at java.net.SocketInputStream.read(SocketInputStream.java:141)
at io.nats.client.impl.SocketDataPort.read(SocketDataPort.java:100)
at io.nats.client.impl.NatsConnectionReader.run(NatsConnectionReader.java:111)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
2019-04-17T09:30:12.413 [pool-4-thread-1] INFO c.b.o.c.nats.NatsConnectionProvider - NATS: connection event - nats: connection disconnected
2019-04-17T09:30:12.419 [pool-4-thread-1] INFO c.b.o.c.nats.NatsConnectionProvider - NATS: connection event - nats: discovered servers
2019-04-17T09:30:12.525 [pool-4-thread-1] ERROR c.b.o.c.nats.NatsConnectionProvider - Nats 'null' exception: {}
java.util.concurrent.CancellationException: null
at java.util.concurrent.CompletableFuture.cancel(CompletableFuture.java:2263)
at io.nats.client.impl.NatsConnection.cleanUpPongQueue(NatsConnection.java:618)
at io.nats.client.impl.NatsConnection.closeSocketImpl(NatsConnection.java:599)
at io.nats.client.impl.NatsConnection.closeSocket(NatsConnection.java:472)
at io.nats.client.impl.NatsConnection.lambda$handleCommunicationIssue$2(NatsConnection.java:445)
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
2019-04-17T09:30:12.527 [pool-4-thread-1] INFO c.b.o.c.nats.NatsConnectionProvider - NATS: connection event - nats: connection disconnected

**Comment by sasbury** [comment]
Can you attach your connection options, they may not matter, but can't hurt. I don't think i was able to reproduce (and as a result be sure I fixed) so if you can provide any help with a small sample/unit test it would be greatly appreciated.

I am sorry this is happening, hopefully we can find the issue quickly (well quickly moving forward clearly i didn't find it quickly yet)

**Comment by igoberman** [comment]
Stephen, thanks for the quick response. I did not even try to duplicate it with unit test as it happens very seldom. But if it does, it is a big problem. The app was very busy at that time and nats server booted client out because 'write deadline' was over default of 2s. I have increased write_deadline to 10s and it did not happen since.
Do you think it is connected somehow to the exception in the 'cleanUpPongQueue'? Or it is unrelated.
I will spend some time trying to reproduce it...

Connection code is below:
private ErrorListener errorListener = new ErrorListener() {

	@Override
	public void slowConsumerDetected(Connection conn, Consumer consumer) {
		logger.error("{}: Slow Consumer {}", conn.getConnectedUrl(), consumer);
	}

	@Override
	public void exceptionOccurred(Connection conn, Exception exp) {
		logger.error("Nats '{}' exception: {}", conn.getConnectedUrl(), exp);
	}

	@Override
	public void errorOccurred(Connection conn, String error) {
		logger.error("Nats '{}': Error {}", conn.getConnectedUrl(), error);
	}
};

Options.Builder natsOptions = new Options.Builder().servers(url)
		.connectionName(config.getAppName());
natsOptions.maxReconnects(-1).reconnectWait(Duration.ofSeconds(1))
		.connectionTimeout(Duration.ofSeconds(5));
natsOptions.pingInterval(Duration.ofSeconds(20));
natsOptions.connectionListener((conn, e) ->
	logger.info("NATS: connection event - {}, connected url: {}. servers: {} ", e, conn.getConnectedUrl(), conn.getServers())
);
natsOptions.errorListener(errorListener);

// Do not cache any messages when Nats connection is down.
natsOptions.reconnectBufferSize(-1);
//
Connection connection = Nats.connect(natsOptions.build());

**Comment by sasbury** [comment]
I wonder if I just have a simple missed catch or missed clean up for the exception during connect. I will take a look. Interesting that you say it happens on a write deadline, maybe i can reproduce it with that.

**Comment by igoberman** [comment]
I was able to duplicate something similar, but not quite (see attachment below). I have both producer and consumer in in one test. Both of them are detected as 'slow consumers'. It runs for couple minutes and then just kind of pauses and sits without any output. It then throws OutOfMemoryError, that is not a good sign.

[ReconnectBug.txt](https://github.com/nats-io/java-nats/files/3109346/ReconnectBug.txt)

Tid: 15, IN, NATS: connection event - nats: connection opened, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: connection opened, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 1, Published 1 messages.
Tid: 21, Received 1 messages
Tid: 21, Received 1000001 messages
Tid: 21, Received 2000001 messages
Tid: 21, Received 3000001 messages
Tid: 15, IN, nats://127.0.0.1:4222: Slow Consumer
Tid: 15, IN, nats://127.0.0.1:4222: Slow Consumer
Tid: 21, Received 4000001 messages
Tid: 15, IN, nats://127.0.0.1:4222: Slow Consumer
Tid: 15, IN, nats://127.0.0.1:4222: Slow Consumer
Tid: 21, Received 5000001 messages
Tid: 20, OUT, Nats 'nats://127.0.0.1:4222' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 20, OUT, NATS: connection event - nats: connection disconnected, connected url: null. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: connection reconnected, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: subscriptions re-established, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 21, Received 6000001 messages
Tid: 20, OUT, Nats 'nats://127.0.0.1:4222' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 21, Received 7000001 messages
Tid: 20, OUT, NATS: connection event - nats: connection disconnected, connected url: null. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: connection reconnected, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: subscriptions re-established, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 21, Received 8000001 messages
Tid: 20, OUT, Nats 'nats://127.0.0.1:4222' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 1, Published 10000001 messages.
Tid: 20, OUT, NATS: connection event - nats: connection disconnected, connected url: null. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: connection reconnected, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: subscriptions re-established, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 15, IN, nats://127.0.0.1:4222: Slow Consumer
Tid: 21, Received 9000001 messages
Tid: 20, OUT, Nats 'nats://127.0.0.1:4222' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 20, OUT, NATS: connection event - nats: connection disconnected, connected url: null. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: connection reconnected, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: subscriptions re-established, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 21, Received 10000001 messages
Tid: 20, OUT, Nats 'nats://127.0.0.1:4222' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 20, OUT, NATS: connection event - nats: connection disconnected, connected url: null. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: connection reconnected, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: subscriptions re-established, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 21, Received 11000001 messages
Tid: 20, OUT, Nats 'nats://127.0.0.1:4222' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 1, Published 20000001 messages.
Tid: 20, OUT, NATS: connection event - nats: connection disconnected, connected url: null. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: connection reconnected, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 20, OUT, NATS: connection event - nats: subscriptions re-established, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 21, Received 12000001 messages
Tid: 20, OUT, Nats 'nats://127.0.0.1:4222' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 15, IN, nats://127.0.0.1:4222: Slow Consumer
Tid: 20, OUT, Nats 'null' exception: java.lang.IllegalStateException: Max outgoing Ping count exceeded.
Tid: 15, IN, nats://127.0.0.1:4222: Slow Consumer
Tid: 15, IN, Nats 'nats://127.0.0.1:4222' exception: java.net.SocketException: Connection reset
Tid: 15, IN, NATS: connection event - nats: connection disconnected, connected url: null. servers: [nats://127.0.0.1:4222] 
Tid: 15, IN, NATS: connection event - nats: connection reconnected, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Tid: 15, IN, NATS: connection event - nats: subscriptions re-established, connected url: nats://127.0.0.1:4222. servers: [nats://127.0.0.1:4222] 
Exception in thread "Nats Connection Timer" Exception in thread "Nats Connection Timer" Exception in thread "pool-2-thread-1" Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
java.lang.OutOfMemoryError: GC overhead limit exceeded

**Comment by sasbury** [comment]
i will try to give it a shot this week, thanks!

**Comment by sasbury** [comment]
this week is turning busy, hopefully i can dig in next week.

**Comment by igoberman** [comment]
Thanks. It could be that they whole issue is that application just runs out of memory if nats cannot keep up. Looks like you use unbounded LinkedBlockingQueue for publication. The size of the queue should be managed, but it looks like it is not. What do you think about adding an option to limit the queue size/ dropping the queue on disconnect?

I found even more serious issue: after reconnect messages are delivered out of sequence. Looks like messages arriving after reconnect are mixed up with queued messages before reconnect. I am attaching another test app.

[NatsReconnectBug.txt](https://github.com/nats-io/java-nats/files/3135510/NatsReconnectBug.txt)

**Comment by sasbury** [comment]
Thanks for the replication test, i did notice you put

        // Do not cache any messages when Nats connection is down.
        natsOptions.reconnectBufferSize(-1);

-1 will cache as much as possible, use 0 to stop caching. Might have been an old comment or something. Regardless the program does blow up.

I am really weirded out by the out of order deliveries, so digging in.

**Comment by sasbury** [comment]
Alright, i was mostly worried about the out of order stuff, and it turns out that was the issue. During a reconnect the library filters the outgoing queue for old ping requests. However, that wasn't happening within a lock so publish message where getting written out of order with the filtered queue. I fixed this (with a small performance price paid) and was able to run the test program (which I also added to the repo ;-) ) for 30M messages and numerous reconnects without any issue.

Thanks again for sticking with this. As I mentioned in the other issue, I don't think we can do anything about the short ping interval, but I think the library is more stable now that we fixed the reconnect issue it was triggering.

I did reduce ping traffic on a subscriber by pushing the ping forward if a message was received. This doesn't help publish-only apps, so in those cases the ping needs to have time to get to the server and back in the normal traffic flow. I added some doc.

The changes are in the 2.4.4 branch if you want to try them before a release. I am hoping to get a release out this week.
